File: browser.go
package main

import (
	"golang.org/x/net/websocket"

	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"io/ioutil"
	"math"
	"net/http"
	"net/url"
	"os"
	"os/exec"
	"strings"
	"sync"
	"syscall"
	"time"
)

type Browser struct {
	cmd    *exec.Cmd
	done   chan bool

	ws     *websocket.Conn

	mux    sync.Mutex

	msgs   []message
	msgidx int             // Write index
	msgcnt int             // Item count 
	msgmux sync.Mutex

	reqid  int
}

// Messages are dynamic json structures divided into responses and events
// Ie only a subset of the following fields will be populated at once
type message struct {
	Received time.Time

	//
	// Response fields
	//

	Id int `json:"id"`
	Result struct {
		Result struct {
			Type  string      `json:"type"`
			Value interface{} `json:"value"`
		} `json:"result"`

		FrameId string `json:"frameId"`
	} `json:"result"`

	//
	// Event fields
	//

	Method string `json:"method"`
	Params struct {
		Frame struct {
			Id string `json:id"`
		} `json:"frame"`
	} `json:"params"`
}

func NewBrowser(incognito bool) (*Browser, error) {
	//
	// Launch headless browser
	//
	
	browser := Browser{
		msgs: make([]message, 1000),
	}

	cmd := []string{"/opt/google/chrome/chrome", "--headless", "--remote-debugging-port=0"}

	if home, ok := os.LookupEnv("HOME"); !ok {
		return nil, errors.New("HOME not defined")
	} else {
		cmd = append(cmd, fmt.Sprintf("--user-data-dir=%s/.config/google-chrome", home))
	}

	if incognito {
		cmd = append(cmd, "--incognito")
	}


	browser.cmd = exec.Command(cmd[0], cmd[1:]...)
	stdout, err := browser.cmd.StdoutPipe()
	if err != nil {
		return nil, err
	}
	stderr, err := browser.cmd.StderrPipe()
	if err != nil {
		return nil, err
	}
	if err := browser.cmd.Start(); err != nil {
		return nil, err
	}
	InfoLog("browser process launched (pid %d)\n", browser.cmd.Process.Pid)

	browser.done = make(chan bool)

	go func() {
		if err := browser.cmd.Wait(); err != nil {
			if exiterr, ok := err.(*exec.ExitError); ok {
				if status, ok := exiterr.Sys().(syscall.WaitStatus); ok {
					if status.Exited() {
						InfoLog("cmd.Wait(): exited with return code %d", status.ExitStatus())
					} else if status.Signaled() {
						InfoLog("cmd.Wait(): %s (%d) signaled", status.Signal().String(), status.Signal())
					} else {
						ErrorLog("cmd.Wait(): unhandled status: %s", exiterr.Error())
					}
				} else {
					ErrorLog("cmd.Wait(): unhandled type: %s", exiterr.Error())
				}
			} else {
				ErrorLog("cmd.Wait(): %s\n", err.Error())
			}
		}
		browser.done <- true
	}()

	//
	// Retrieve and connect to the debugger url (via a websocket)
	// Also process stdout/stderr output (necessary to retrieve the debugger url)
	//

	devtools := make(chan string)

	for index, reader := range []io.ReadCloser{stdout, stderr} {
		go func() {
			sent := false
			output := func(line string) {
				if len(line) > 0 {
					DebugLog("browser/%d: %s", index + 1, line)

					if index == 1 && !sent {
						prefix := "DevTools listening on "
						if strings.HasPrefix(line, prefix) {
							devtools <- line[len(prefix):]
							close(devtools)
							sent = true
						}
					}
				}
			}

			var linebuf bytes.Buffer
			buf := make([]byte, 1024)

			for {
				n, err := reader.Read(buf)
				if err != nil {
					if !(errors.Is(err, io.EOF) || errors.Is(err, os.ErrClosed)) {
						ErrorLog("<%d>.Read(): %s", index + 1, err.Error())
					}
					break
				}
				if n > 0 {
					splits := bytes.SplitAfter(buf[:n], []byte{'\n'})
					if len(splits) == 1 {
						linebuf.Write(buf[:n])
					} else {
						output(linebuf.String() + strings.TrimSuffix(string(splits[0]), "\n"))
						for i := 1; i < len(splits) - 1; i++ {
							output(strings.TrimSuffix(string(splits[i]), "\n"))
						}

						linebuf.Reset()
						linebuf.Write(splits[len(splits) - 1])
					}
				}
			}
		}()
	}

	var devtoolsUrl string

	select {
	case <- browser.done:
		return nil, errors.New("browser exited early")
	case devtoolsUrl = <- devtools:
		DebugLog("devtools url: %s", devtoolsUrl)
	case <- time.After(15 * time.Second):
		browser.Close()
		return nil, errors.New("deadline reached looking for devtools url")
	}

	parsedUrl, err := url.Parse(devtoolsUrl)
	if err != nil {
		browser.Close()
		return nil, err
	}

	parsedUrl.Scheme = "http"
	parsedUrl.Path = "/json"

	time.Sleep(5 * time.Second)

	client := http.Client{ Timeout: 15 * time.Second }
	req, err := http.NewRequest("GET", parsedUrl.String(), nil)
	if err != nil {
		browser.Close()
		return nil, err
	}
	resp, err := client.Do(req)
	if err != nil {
		browser.Close()
		return nil, err
	}

	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		browser.Close()
		return nil, fmt.Errorf("bad status code (%d) from %s", resp.StatusCode, parsedUrl.String())
	}

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		browser.Close()
		return nil, err
	}

	type entry struct {
		DebuggerUrl string `json:"webSocketDebuggerUrl"`
	}

	var entries []entry
	if err := json.Unmarshal(body, &entries); err != nil {
		browser.Close()
		return nil, err
	}

	if len(entries) != 1 {
		browser.Close()
		return nil, fmt.Errorf("unexpected entry count (%d)", len(entries))
	}
	debuggerUrl := entries[0].DebuggerUrl
	DebugLog("debugger url: %s", debuggerUrl)

	browser.ws, err = websocket.Dial(debuggerUrl, "", "http://127.0.0.1/")
	if err != nil {
		browser.Close()
		return nil, err
	}

	//
	// Define a goroutine to process websocket messages
	// To be held in a FIFO circular buffer
	//
	
	go func() {
		var msgbuf bytes.Buffer
		chunk := make([]byte, 4096)

		for {
			n, err := browser.ws.Read(chunk)
			if err != nil {
				//if !errors.Is(err, net.ErrClosed) {
				if !(strings.HasSuffix(err.Error(), "use of closed network connection") || errors.Is(err, io.EOF)) {
					ErrorLog(err)
				}
				break
			}
			if n > 0 {
				var msg message

				msgbuf.Write(chunk[:n])
				if err := json.Unmarshal(msgbuf.Bytes(), &msg); err != nil {
					continue
				}

				TraceLog("<<< %s", msgbuf.String())
				msgbuf.Reset()

				// Not all events include a timestamp param
				msg.Received = time.Now()

				browser.msgmux.Lock()

				browser.msgs[browser.msgidx] = msg

				browser.msgidx = (browser.msgidx + 1) % len(browser.msgs)
				if browser.msgcnt < len(browser.msgs) {
					browser.msgcnt++
				}

				browser.msgmux.Unlock()
			}
		}
	}()

	//
	// Debugger initialization
	//

	timeout := 5 * time.Second

	useragent, err := browser.EvaluateString("navigator.userAgent", nil)
	if err != nil {
		browser.Close()
		return nil, err
	}
	useragent = strings.Replace(useragent, "HeadlessChrome", "Chrome", 1)
	if _, err := browser.execute("Network.setUserAgentOverride", map[string]interface{}{ "userAgent": useragent }, timeout); err != nil {
		browser.Close()
		return nil, err
	}

	// Enable notification of Page domain events
	// TODO Consider also enabling notification of DOM and Network events
	if _, err := browser.execute("Page.enable", nil, timeout); err != nil {
		browser.Close()
		return nil, err
	}

	return &browser, nil
}

func (b *Browser) Close() error {
	b.mux.Lock()
	defer b.mux.Unlock()

	done := false
	if b.ws != nil {
		if _, err := b.execute("Browser.close", nil, -1); err != nil {
			ErrorLog(err)
		}

		select {
		case <- b.done:
			done = true
		case <- time.After(15 * time.Second):
			WarningLog("deadline reached waiting for Browser.close")
		}

		if err := b.ws.Close(); err != nil {
			ErrorLog(err)
		}
	}

	if !done {
		if err := b.cmd.Process.Signal(os.Interrupt); err != nil {
			return err
		}

		select {
		case <- b.done:
			done = true
		case <- time.After(15 * time.Second):
			WarningLog("deadline reached waiting for SIGINT")
			if err := b.cmd.Process.Kill(); err != nil {
				return err
			}
		}

	}

	return nil
}

func (b *Browser) waitfor(predicate func(m *message) bool, timeout time.Duration) (*message, error) {
	stop := time.Now().Add(timeout)
	for {
		if timeout > 0 && time.Now().After(stop) {
			return nil, nil
		}

		b.msgmux.Lock()

		if b.msgcnt > 0 {
			var i, j int
			if b.msgcnt == len(b.msgs) {
				i = b.msgidx
				j = (b.msgidx + 1) % len(b.msgs)
			} else {
				i = 0
				j = b.msgcnt - 1
			}

			k := i
			for {
				if predicate(&b.msgs[k]) {
					msg := b.msgs[k]
					b.msgmux.Unlock()
					return &msg, nil
				}

				if k == j {
					break
				}
				k = (k + 1) % len(b.msgs)
			}
		}

		b.msgmux.Unlock()

		time.Sleep(100 * time.Millisecond)
	}
}

func (b *Browser) execute(method string, params map[string]interface{}, timeout time.Duration) (*message, error) {
	req := struct {
		Id     int                    `json:"id"`
		Method string                 `json:"method"`
		Params map[string]interface{} `json:"params,omitempty"`
	}{
		Id:     b.reqid,
		Method: method,
		Params: params,
	}
	b.reqid++

	reqbytes, err := json.Marshal(req)
	if err != nil {
		return nil, err
	}

	TraceLog(">>> %s", reqbytes)

	if _, err = b.ws.Write(reqbytes); err != nil {
		return nil, err
	}

	// timeout < 0 => don't wait for a response
	// timeout = 0 => wait for a response indefinitely
	// timeout > 0 => wait until (at most) timeout

	if timeout < 0 {
		return nil, nil
	}

	if msg, err := b.waitfor(func(m *message) bool { return m.Id == req.Id }, timeout); err != nil {
		return nil, err
	} else {
		if msg == nil {
			return nil, errors.New("execution timed out")
		} else {
			return msg, nil
		}
	}
}

func (b *Browser) Evaluate(expr string, options map[string]interface{}) error {
	b.mux.Lock()
	defer b.mux.Unlock()

	timeout := 5 * time.Second
	if t, ok := options["timeout"].(time.Duration); ok {
		timeout = t
	}

	_, err := b.execute("Runtime.evaluate", map[string]interface{}{ "expression": expr }, timeout)
	return err
}

func (b *Browser) EvaluateBool(expr string, options map[string]interface{}) (bool, error) {
	b.mux.Lock()
	defer b.mux.Unlock()

	timeout := 5 * time.Second
	if t, ok := options["timeout"].(time.Duration); ok {
		timeout = t
	}

	msg, err := b.execute("Runtime.evaluate", map[string]interface{}{ "expression": expr }, timeout)
	if err != nil {
		return false, err
	}

	if msg.Result.Result.Type != "boolean" {
		return false, fmt.Errorf("unexpected type: %s", msg.Result.Result.Type)
	}
	if value, ok := msg.Result.Result.Value.(bool); !ok {
		return false, errors.New("value type assertion failed")
	} else {
		return value, nil
	}
}

func (b *Browser) EvaluateInt(expr string, options map[string]interface{}) (int, error) {
	b.mux.Lock()
	defer b.mux.Unlock()

	timeout := 5 * time.Second
	if t, ok := options["timeout"].(time.Duration); ok {
		timeout = t
	}

	msg, err := b.execute("Runtime.evaluate", map[string]interface{}{ "expression": expr }, timeout)
	if err != nil {
		return 0, err
	}

	if msg.Result.Result.Type != "number" {
		return 0, fmt.Errorf("unexpected type: %s", msg.Result.Result.Type)
	}
	if value, ok := msg.Result.Result.Value.(float64); !ok {
		return 0, errors.New("value type assertion failed")
	} else {
		if math.Floor(value) != value {
			return 0, errors.New("value type check failed")
		} else {
			return int(value), nil
		}
	}
}

func (b *Browser) EvaluateFloat(expr string, options map[string]interface{}) (float64, error) {
	b.mux.Lock()
	defer b.mux.Unlock()

	timeout := 5 * time.Second
	if t, ok := options["timeout"].(time.Duration); ok {
		timeout = t
	}

	msg, err := b.execute("Runtime.evaluate", map[string]interface{}{ "expression": expr }, timeout)
	if err != nil {
		return 0.0, err
	}

	if msg.Result.Result.Type != "number" {
		return 0.0, fmt.Errorf("unexpected type: %s", msg.Result.Result.Type)
	}
	if value, ok := msg.Result.Result.Value.(float64); !ok {
		return 0.0, errors.New("value type assertion failed")
	} else {
		return value, nil
	}
}

func (b *Browser) EvaluateString(expr string, options map[string]interface{}) (string, error) {
	b.mux.Lock()
	defer b.mux.Unlock()

	timeout := 5 * time.Second
	if t, ok := options["timeout"].(time.Duration); ok {
		timeout = t
	}

	msg, err := b.execute("Runtime.evaluate", map[string]interface{}{ "expression": expr }, timeout)
	if err != nil {
		return "", err
	}

	if msg.Result.Result.Type != "string" {
		return "", fmt.Errorf("unexpected type: %s", msg.Result.Result.Type)
	}
	if value, ok := msg.Result.Result.Value.(string); !ok {
		return "", errors.New("value type assertion failed")
	} else {
		return value, nil
	}
}

func (b *Browser) Goto(url string, options map[string]interface{}) (bool, error) {
	b.mux.Lock()
	defer b.mux.Unlock()

	cmdTimeout := 5 * time.Second

	gotoTimeout := 60 * time.Second
	if t, ok := options["timeout"].(time.Duration); ok {
		gotoTimeout = t
	}

	var referrer string
	if r, ok := options["referrer"].(string); ok {
		referrer = r
	}

	params := map[string]interface{}{ "url": url }
	if referrer != "" {
		params["referrer"] = referrer
	}

	msg, err := b.execute("Page.navigate", params, cmdTimeout)
	if err != nil {
		return false, err
	}
	if msg.Result.FrameId == "" {
		return false, errors.New("frameId missing")
	}
	DebugLog("%+v", msg.Result.FrameId)

	pred := func(m *message) bool { return m.Method == "Page.frameNavigated" && m.Params.Frame.Id == msg.Result.FrameId }
	if evt, err := b.waitfor(pred, gotoTimeout); err != nil {
		return false, err
	} else if evt == nil {
		return true, nil
	} else {
		return false, nil
	}
}

// TODO Add support for simultaneous sessions

File: src/browser.rs
use nix::sys::signal::{kill, Signal};
use nix::unistd::Pid;

use regex::Regex;

use reqwest::{self, StatusCode};

use serde::{Deserialize, Serialize};

use serde_json::{self, Map, Number, Value};

use std::env;
use std::error::Error;
use std::fmt::{self, Debug, Formatter};
use std::io::{self, BufRead, BufReader};
use std::net::TcpStream;
use std::process::{Child, Command, Stdio};
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::{Duration, Instant};

use timeout_readwrite::TimeoutReader;

use tungstenite::{self, protocol::{self, WebSocket}};

use url::Url;

#[derive(Clone, Debug, Deserialize)]
struct ResponseResultResult {
    #[serde(rename = "type")]
    type_: String,
    value: Value,
}

#[derive(Clone, Debug, Deserialize)]
struct ResponseResult {
    result: Option<ResponseResultResult>,
    #[serde(rename = "frameId")]
    frame_id: Option<String>,
    #[serde(rename = "loaderId")]
    loader_id: Option<String>,
}

#[derive(Clone, Debug, Deserialize)]
struct Response {
    id: u64,
    result: ResponseResult,
}

#[derive(Debug, Deserialize)]
struct EventParamsFrame {
    id: String,
    #[serde(rename = "loaderId")]
    loader_id: String,
}

#[derive(Debug, Deserialize)]
struct EventParams {
    frame: EventParamsFrame,
}

#[derive(Debug, Deserialize)]
struct Event {
    method: String,
    params: EventParams,
}

#[derive(Debug, Deserialize)]
#[serde(untagged)]
enum Message {
    Response(Response),
    Event(Event),
}

struct BrowserShared {
    websocket: WebSocket<TcpStream>,
    ringbuf: RingBuf<Message>,
}

pub struct Browser {
    child: Child,
    shared: Arc<Mutex<BrowserShared>>,
    request_id: u64,
}

impl Browser {
    pub fn new() -> Result<Self, Box<dyn Error>> {
        let mut cmd = Command::new("/opt/google/chrome/chrome");
        cmd.args(&["--headless", "--remote-debugging-port=0"]);

        match env::var("HOME") {
            Ok(h) => { cmd.arg(format!("--user-data-dir={}/.config/google-chrome", h)); },
            Err(e) => return Err(Box::new(e))
        }

        cmd.stdin(Stdio::null())
           .stdout(Stdio::null())
           .stderr(Stdio::piped());

        let mut child = match cmd.spawn() {
            Ok(c) => c,
            Err(e) => return Err(Box::new(e))
        };

        let stderr = child.stderr.take().unwrap();

        // Ideally would unwrap stderr from the BufReader and TimeoutReader for later use in the thread below.
        // However (at least currently) only BufReader supports into_inner() to unwrap the underlying reader.
        let mut stderr_reader = BufReader::new(TimeoutReader::new(stderr, Duration::from_secs(1)));

        let mut port: Option<u16> = None;
        {
            let re = Regex::new(r"^DevTools listening on (ws://127.0.0.1:(\d+)/devtools/browser/[a-f0-9-]+)$").unwrap();

            for _ in 0..15 {
                let mut line = String::new();
                match stderr_reader.read_line(&mut line) {
                    Ok(_) => {
                        if line.ends_with('\n') {
                            line.pop();
                        }
                        if !line.is_empty() {
                            log::info!("stderr: {}", line);

                            if let Some(c) = re.captures(&line) {
                                log::info!("websocket url: {}", c.get(1).unwrap().as_str());

                                let s = c.get(2).unwrap().as_str();
                                match s.parse::<u16>() {
                                    Ok(p) => {
                                        port = Some(p);
                                        break;
                                    },
                                    Err(e) => log::error!("error parsing port {}: {}", s, e)
                                }
                            }
                        }
                    },
                    Err(e) if e.kind() == io::ErrorKind::TimedOut => {},
                    Err(e) => {
                        log::error!("error reading stderr: {}", e);
                        break;
                    }
                }
            }

            if let None = port {
                let _ = child.kill();
                return Err(string_error::static_err("websocket url not found"));
            }
        }
        let port = port.unwrap();

        let debugger_url: String;
        {
            #[derive(Debug, Deserialize)]
            struct Config {
                #[serde(rename = "webSocketDebuggerUrl")]
                debugger_url: String,
            }

            match reqwest::blocking::get(format!("http://127.0.0.1:{}/json", port)) {
                Ok(r) if r.status() == StatusCode::OK => {
                    match r.json::<Vec<Config>>() {
                        Ok(c) if c.len() == 1 => {
                            debugger_url = c[0].debugger_url.clone();
                            log::info!("debugger url: {}", debugger_url);
                        },
                        Ok(_) => {
                            let _ = child.kill();
                            return Err(string_error::static_err("unexpected json url response config count"));
                        },
                        Err(e) => {
                            let _ = child.kill();
                            return Err(Box::new(e));
                        }
                    }
                },
                Ok(r) => {
                    let _ = child.kill();
                    return Err(string_error::into_err(format!("{} status from json url", r.status())));
                },
                Err(e) => {
                    let _ = child.kill();
                    return Err(Box::new(e));
                }
            }
        }

        thread::spawn(move || {
            for line in stderr_reader.lines() {
                match line {
                    Ok(l) => {
                        if !l.is_empty() {
                            log::info!("stderr: {}", l);
                        }
                    },
                    Err(e) if e.kind() == io::ErrorKind::TimedOut => {},
                    Err(e) => {
                        log::error!("error reading stderr: {}", e);
                        break;
                    }
                }
            }
        });

        // Providing a non-blocking stream so that websocket will support non-blocking reads.
        // Ref: https://docs.rs/tungstenite/0.17.2/tungstenite/client/fn.client.html
        //      https://doc.rust-lang.org/stable/std/net/struct.TcpStream.html#method.set_nonblocking
        let stream: TcpStream;
        {
            let port = match Url::parse(&debugger_url) {
                Ok(u) => u.port().unwrap_or(80),
                Err(e) => {
                    let _ = child.kill();
                    return Err(Box::new(e));
                }
            };

            stream = match TcpStream::connect(format!("127.0.0.1:{}", port)) {
                Ok(s) => s,
                Err(e) => {
                    let _ = child.kill();
                    return Err(Box::new(e));
                }
            };
        }

        let mut websocket: WebSocket<TcpStream>;
        match tungstenite::client::client(debugger_url, stream) {
            Ok((w, r)) if r.status() == StatusCode::SWITCHING_PROTOCOLS => websocket = w,
            Ok((_, r)) => {
                let _ = child.kill();
                return Err(string_error::into_err(format!("{} status from debugger url", r.status())));
            },
            Err(e) => {
                let _ = child.kill();
                return Err(Box::new(e));
            }
        }

        if let Err(e) = websocket.get_mut().set_nonblocking(true) {
            let _ = child.kill();
            return Err(Box::new(e));
        }

        let shared = Arc::new(Mutex::new(BrowserShared{
            websocket: websocket,
            ringbuf: RingBuf::<Message>::new(),
        }));

        {
            let shared = shared.clone();
            thread::spawn(move || {
                loop {
                    match shared.lock() {
                        Ok(ref mut s) => {
                            match s.websocket.read_message() {
                                Ok(tungstenite::protocol::Message::Text(m)) => {
                                    log::trace!("<<< {}", m);
                                    match serde_json::from_str::<Message>(&m) {
                                        Ok(m) => s.ringbuf.push(m),
                                        Err(_e) => {
                                            // Only a single event format is defined by Message::Event enum.
                                            // Add more (ideally all possible) for thie error message to be useful.
                                            // log::error!("json deserialization error: {:?}", _e);
                                        }
                                    }
                                },
                                Ok(_) => log::warn!("received non-text websocket message"),
                                Err(tungstenite::error::Error::Io(e)) if e.kind() == io::ErrorKind::WouldBlock => {},
                                Err(tungstenite::error::Error::Protocol(tungstenite::error::ProtocolError::ResetWithoutClosingHandshake)) => break,
                                Err(e) => {
                                    log::error!("websocket error: {}", e);
                                    break;
                                }
                            }
                        },
                        Err(e) => {
                            log::error!("poisoned mutex: {}", e);
                            break;
                        }
                    }
                    thread::sleep(Duration::from_millis(100));
                }
            });
        }

        let mut retval = Self {
            child: child,
            shared: shared,
            request_id: 0,
        };

        {
            let mut user_agent = match retval.evaluate_string("navigator.userAgent", Some(Duration::from_secs(5))) {
                Ok(s) => s,
                Err(e) => {
                    drop(retval);
                    return Err(e);
                }
            };
            user_agent = user_agent.replacen("HeadlessChrome", "Chrome", 1);

            let mut params: Map<String, Value> = Map::new();
            params.insert("userAgent".to_string(), Value::String(user_agent.to_string()));
            if let Err(e) = retval.execute("Network.setUserAgentOverride", Some(params), Some(Duration::from_secs(5))) {
                drop(retval);
                return Err(e);
            }

            if let Err(e) = retval.execute("Page.enable", None, Some(Duration::from_secs(5))) {
                drop(retval);
                return Err(e);
            }
        }

        Ok(retval)
    }

    fn execute(&mut self, method: &str, params: Option<Map<String, Value>>, timeout: Option<Duration>) -> Result<Response, Box<dyn Error>> {
        #[derive(Serialize)]
        struct Request {
            id: Number,
            method: String,
            #[serde(skip_serializing_if = "Option::is_none")]
            params: Option<Map<String, Value>>,
        }

        let request = Request{
            id: Number::from(self.request_id),
            method: method.to_string(),
            params: params,
        };
        self.request_id += 1;

        let shared = self.shared.clone();

        {
            let request_string = match serde_json::to_string(&request) {
                Ok(s) => s,
                Err(e) => return Err(Box::new(e))
            };
            log::trace!(">>> {}", request_string);

            match shared.lock() {
                Ok(mut s) => {
                    match s.send_request(&request_string) {
                        Ok(_) => {},
                        Err(e) => return Err(e)
                    }
                },
                Err(e) => return Err(string_error::into_err(format!("poisoned mutex: {}", e)))
            }
        }

        {
            let start = Instant::now();
            loop {
                if let Some(t) = timeout {
                    if Instant::now().duration_since(start) > t {
                        return Err(Box::new(io::Error::new(io::ErrorKind::TimedOut, "timed out")));
                    }
                }

                match shared.lock() {
                    Ok(s) => {
                        for message in s.ringbuf.iter() {
                            if let Message::Response(r) = message {
                                if let Some(i) = request.id.as_u64() {
                                    if r.id == i {
                                        return Ok(r.clone());
                                    }
                                }
                            }
                        }
                    },
                    Err(e) => return Err(string_error::into_err(format!("poisoned mutex: {}", e)))
                }

                thread::sleep(Duration::from_millis(100));
            }
        }
    }

    pub fn evaluate_string(&mut self, expr: &str, timeout: Option<Duration>) -> Result<String, Box<dyn Error>> {
        let mut params: Map<String, Value> = Map::new();
        params.insert("expression".to_string(), Value::String(expr.to_string()));
        match self.execute("Runtime.evaluate", Some(params), timeout) {
            Ok(r) => {
                if let Some(s) = r.result.result {
                    if s.type_ == "string" {
                        Ok(s.value.to_string())
                    } else {
                        Err(string_error::into_err(format!("response result wrong type: {:?}", s.type_)))
                    }
                } else {
                    Err(string_error::into_err(format!("response missing result: {:?}", r)))
                }
            },
            Err(e) => Err(e)
        }
    }

    pub fn goto(&mut self, url: &str, timeout: Option<Duration>, referrer: Option<&str>) -> Result<(), Box<dyn Error>> {
        let mut params: Map<String, Value> = Map::new();
        params.insert("url".to_string(), Value::String(url.to_string()));
        if let Some(r) = referrer {
            params.insert("referrer".to_string(), Value::String(r.to_string()));
        }

        let (frame_id, loader_id): (String, String);
        match self.execute("Page.navigate", Some(params), Some(Duration::from_secs(5))) {
            Ok(r) => {
                if let Some(s) = r.result.frame_id {
                    frame_id = s;
                } else {
                    return Err(string_error::static_err("response result missing frame_id"));
                }
                if let Some(s) = r.result.loader_id {
                    loader_id = s;
                } else {
                    return Err(string_error::static_err("response result missing loader_id"));
                }
            },
            Err(e) => return Err(e)
        }

        let shared = self.shared.clone();
        {
            let start = Instant::now();
            loop {
                if let Some(t) = timeout {
                    if Instant::now().duration_since(start) > t {
                        return Err(Box::new(io::Error::new(io::ErrorKind::TimedOut, "timed out")));
                    }
                }

                match shared.lock() {
                    Ok(s) => {
                        for message in s.ringbuf.iter() {
                            if let Message::Event(e) = message {
                                if e.method == "Page.frameNavigated" && e.params.frame.id == frame_id && e.params.frame.loader_id == loader_id {
                                    return Ok(());
                                }
                            }
                        }
                    },
                    Err(e) => return Err(string_error::into_err(format!("poisoned mutex: {}", e)))
                }

                thread::sleep(Duration::from_millis(100));
            }
        }
    }
}

impl Drop for Browser {
    fn drop(&mut self) {
        match self.execute("Browser.close", None, Some(Duration::from_secs(0))) {
            Ok(_) => {},
            Err(e) => {
                if let Some(f) = e.downcast_ref::<io::Error>() {
                    if f.kind() != io::ErrorKind::TimedOut {
                        log::error!("Browser.close error: {}", f);
                    }
                } else {
                    log::error!("Browser.close error: {}", e);
                }
            }
        }

        let mut alive = true;
        let start = Instant::now();
        loop {
            match self.child.try_wait() {
                Ok(Some(_)) => {
                    alive = false;
                    break;
                },
                Ok(None) => {},
                Err(e) => {
                    log::error!("process try_wait error: {}", e);
                    break;
                }
            }

            if Instant::now().duration_since(start) > Duration::from_secs(15) {
                break;
            }
            thread::sleep(Duration::from_millis(100));
        }


        if !alive {
            return;
        }
        log::warn!("deadline reached waiting for Browser.close");

        match kill(Pid::from_raw(self.child.id() as i32), Signal::SIGTERM) {
            Ok(()) => {
                for _ in 0..15 {
                    match self.child.try_wait() {
                        Ok(Some(_)) => {
                            alive = false;
                            break;
                        },
                        Ok(None) => {},
                        Err(e) => {
                            log::error!("process try_wait error: {}", e);
                            break;
                        }
                    }
                    thread::sleep(Duration::from_secs(1));
                }
            },
            Err(e) => log::error!("error sending SIGTERM: {}", e)
        }

        if !alive {
            return;
        }
        log::warn!("deadline reached waiting for SIGTERM");

        if let Err(e) = self.child.kill() {
            log::error!("process kill error: {}", e);
        }
    }
}

impl Debug for Browser {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        write!(f, "Browser {{ pid: {} }}", self.child.id())
    }
}

impl BrowserShared {
    fn send_request(&mut self, request: &str) -> Result<(), Box<dyn Error>> {
        match self.websocket.write_message(protocol::Message::Text(request.to_string())) {
            Ok(_) => Ok(()),
            Err(e) => Err(Box::new(e))
        }
    }
}

struct RingBuf<T> {
    buffer: Vec<T>,
    write_idx: usize,
}

struct RingBufIter<'a, T> {
    ringbuf: &'a RingBuf<T>,
    stop_idx: usize,
    read_idx: usize,
    done: bool,
}

impl<T> RingBuf<T> {
    fn new() -> Self {
        Self {
            buffer: Vec::with_capacity(1000),
            write_idx: 0,
        }
    }

    fn push(&mut self, value: T) {
        if self.buffer.len() < self.buffer.capacity() {
            self.buffer.push(value);
        } else {
            self.buffer[self.write_idx] = value;
            self.write_idx = (self.write_idx + 1) % self.buffer.capacity();
        }
    }

    fn iter(&self) -> RingBufIter<T> {
        let (done, start_idx, stop_idx) = if self.buffer.len() == 0 {
            (true, 0, 0)
        } else if self.buffer.len() < self.buffer.capacity() {
            (false, 0, self.buffer.len())
        } else {
            (false, self.write_idx, self.write_idx)
        };

        RingBufIter{
            ringbuf: self,
            stop_idx: stop_idx,
            read_idx: start_idx,
            done: done,
        }
    }
}

impl<'a, T> Iterator for RingBufIter<'a, T> {
    type Item = &'a T;

    fn next(&mut self) -> Option<Self::Item> {
        if self.done {
            None
        } else {
            let retval = Some(&self.ringbuf.buffer[self.read_idx]);
            self.read_idx = (self.read_idx + 1) % self.ringbuf.buffer.capacity();
            if self.read_idx == self.stop_idx {
                self.done = true;
            }
            retval
        }
    }
}

// TODO Add Browser.new_incognito()
// TODO Add Browser.evalute_number(), etc as needed
//      Also consider adding Browser.evaluate (no return type) and replace use of execute() in new()
// TODO Add support for thread-safe, simultaneous sessions
